### PR TITLE
Push preliminary AWS deployment documentation

### PR DESCRIPTION
**Pull request by damianavila**
This is the first iteration in the creation of a how-to guide describing how you can manually set up an AWS kops-based cluster and use our deployer to manually deploy a new hub on top of it.

More details in the commit message (and in the document itself) :wink:

Draft PR for now!

### CODE REVIEW

**Comment by damianavila** [comment]
@choldgraf, your wizard-ish documentation wands are required here 😛  .
A lot of thanks in advance!!

Btw, this one, once it is finished should close #453 and #366 (in cascade).

File: docs/howto/operate/add-aws-hub.md
Comment (consideRatio) [inline docs/howto/operate/add-aws-hub.md:48]: What does <2i2c> mean here? Should it be `2i2c-<hub-name>-kops-state`?

Comment (damianavila) [reply docs/howto/operate/add-aws-hub.md:48]: You are totally right, thanks for noticing it!

File: docs/howto/operate/add-aws-hub.md
Comment (yuvipanda) [inline docs/howto/operate/add-aws-hub.md:102]: Add commands for this too?

File: docs/howto/operate/add-aws-hub.md
Comment (yuvipanda) [inline docs/howto/operate/add-aws-hub.md:143]:
```suggestion
1. Install `boto3` with pip or conda
```

File: docs/howto/operate/add-aws-hub.md
Comment (yuvipanda) [inline docs/howto/operate/add-aws-hub.md:148]:
```suggestion
python3 setup-efs.py <cluster_name>hub.k8s.local <region-of-cluster>
```

File: docs/howto/operate/add-aws-hub.md
Comment (choldgraf) [inline docs/howto/operate/add-aws-hub.md:24]:
```suggestion
and configure it to use the provided AWS user credentials. [Follow this guide](https://docs.aws.amazon.com/cli/latest/userguide/cli-configure-quickstart.html#cli-configure-quickstart-config)
for a quick configuration process.
```

File: docs/howto/operate/add-aws-hub.md
Comment (choldgraf) [inline docs/howto/operate/add-aws-hub.md:48]:
```suggestion
export KOPS_STATE_STORE=s3://2i2c-<hub-name>-kops-state
```

File: docs/howto/operate/add-aws-hub.md
Comment (choldgraf) [inline docs/howto/operate/add-aws-hub.md:54]:
```suggestion
Kops [recommends versioning the S3 bucket](https://kops.sigs.k8s.io/getting_started/aws/#cluster-state-storage) in case you ever need to revert or recover a previous state
```

File: docs/howto/operate/add-aws-hub.md
Comment (choldgraf) [inline docs/howto/operate/add-aws-hub.md:66]: I'm not sure what "understand" means in this context

File: docs/howto/operate/add-aws-hub.md
Comment (choldgraf) [inline docs/howto/operate/add-aws-hub.md:74]: Is this file just an unnecessary extra created by the step that has no purpose? Could you add a short sentence noting why this file is OK to be deleted?

File: docs/howto/operate/add-aws-hub.md
Comment (choldgraf) [reply docs/howto/operate/add-aws-hub.md:102]: Or we could add a separate section on encrypting private keys, since that is probably a generic thing and not AWS-specific, and cross-link it?

File: docs/howto/operate/add-aws-hub.md
Comment (choldgraf) [inline docs/howto/operate/add-aws-hub.md:106]:
```suggestion
3. Build the cluster with the following command (notice that you are passing the ssh public key you just 
```

File: docs/howto/operate/add-aws-hub.md
Comment (choldgraf) [inline docs/howto/operate/add-aws-hub.md:129]:
````suggestion
```{seealso}
More details [in this issue](https://github.com/kubernetes/kops/issues/11199)
```
````

File: docs/howto/operate/add-aws-hub.md
Comment (choldgraf) [inline docs/howto/operate/add-aws-hub.md:131]:
```suggestion
1. After the failed validation finished (you can run these commands in another terminal 
```

File: docs/howto/operate/add-aws-hub.md
Comment (choldgraf) [inline docs/howto/operate/add-aws-hub.md:176]:
```suggestion
You can use of the existing cluster config files as a "template" for your cluster (for example, [here is the Farallon Institute config file](https://github.com/2i2c-org/pilot-hubs/blob/master/config/hubs/farallon.cluster.yaml)).
You may need to tweak names, `serverIP` and singleuser's images references. Make sure 
```

**Review by choldgraf — COMMENTED** [review]
This looks great - thanks for documenting all of this @damianavila - I added in a few suggestions etc, just to make the links a bit easier to discover and follow. Perhaps once this is merged @sgibson91 would be interested in giving the instructions a whirl and reporting back where they are unclear :-)

File: docs/howto/operate/add-aws-hub.md
Comment (damianavila) [reply docs/howto/operate/add-aws-hub.md:102]: Both options are OK... I think I would probably add the step here for now (I was thinking in that direction before pushing this one) and then we can add a new section and cross-reference in another PR...

File: docs/howto/operate/add-aws-hub.md
Comment (damianavila) [reply docs/howto/operate/add-aws-hub.md:122]: > Should we just always run the coredns patch first before validating?

Yep, that also crossed through my head before and you validated it with your comment :wink:

File: docs/howto/operate/add-aws-hub.md
Comment (damianavila) [reply docs/howto/operate/add-aws-hub.md]: Thanks!

File: docs/howto/operate/add-aws-hub.md
Comment (damianavila) [reply docs/howto/operate/add-aws-hub.md:148]: Thanks!

File: docs/howto/operate/add-aws-hub.md
Comment (damianavila) [reply docs/howto/operate/add-aws-hub.md:63]: Kops is able to process the incoming config file to create the desired cluster: https://kops.sigs.k8s.io/cli/kops_create/

Comment (damianavila) [reply docs/howto/operate/add-aws-hub.md:63]: Alternative suggestion accepted :wink: .

File: docs/howto/operate/add-aws-hub.md
Comment (damianavila) [reply docs/howto/operate/add-aws-hub.md:71]: I am not being clear enough, I think... the generated yaml file will be used with the `kops create` command... the issue here is that jsonnet creates this file with 3 points at the very end of the file.

File: docs/howto/operate/add-aws-hub.md
Comment (damianavila) [reply docs/howto/operate/add-aws-hub.md:126]: Nice, thanks!

**Comment by damianavila** [comment]
@choldgraf thanks for all those comments and fixes. 
Btw, I have merged some suggestions and replied with answers to your questions.

**Comment by choldgraf** [comment]
Sounds good - this looks good for a merge IMO once @yuvipanda's comment is addressed, we can iterate and improve upon the documentation in chunks as more team members follow these steps.

**Comment by damianavila** [comment]
Great, I will address the remaining comments, get the PR out of the draft state for a few hours and merge by tomorrow if I do not hear anything else from any of you.

File: docs/howto/operate/add-aws-hub.md
Comment (damianavila) [reply docs/howto/operate/add-aws-hub.md:99]: Added a new 4h step to relocate and encrypt the private key.

File: docs/howto/operate/add-aws-hub.md
Comment (damianavila) [reply docs/howto/operate/add-aws-hub.md:119]: OK, asked for validation after applying the patches.

File: docs/howto/operate/add-aws-hub.md
Comment (damianavila) [reply docs/howto/operate/add-aws-hub.md:63]: Resolving the conversation as per 👍

File: docs/howto/operate/add-aws-hub.md
Comment (damianavila) [reply docs/howto/operate/add-aws-hub.md:71]: Resolving since I think the text is clear enough :wink:

**Review by sgibson91 — APPROVED** [review]
Generally looks good to me, in terms of ease of reading and lots of cross-links to other resources 👍🏻 I'd be keen to give it a try on AWS at some point and provide further technical feedback 🤖

File: docs/howto/operate/add-aws-hub.md
Comment (yuvipanda) [inline docs/howto/operate/add-aws-hub.md:36]: I think an easier / safer way is to put these in `~/.aws/credentials` and just export `AWS_PROFILE`. Does `aws configure` populate `~/.aws/credentials`?

File: docs/howto/operate/add-aws-hub.md
Comment (yuvipanda) [inline docs/howto/operate/add-aws-hub.md:31]:
```suggestion
3. Export some helpful AWS environment variables (because `aws configure` doesn't export
```

File: docs/howto/operate/add-aws-hub.md
Comment (yuvipanda) [inline docs/howto/operate/add-aws-hub.md:47]:
```suggestion
export KOPS_STATE_STORE=s3://2i2c-<cluster-name>-kops-state
```

Where `cluster-name` also matches what's under `config/hubs/<cluster-name>.cluster.yaml`?

File: docs/howto/operate/add-aws-hub.md
Comment (yuvipanda) [inline docs/howto/operate/add-aws-hub.md:48]:
```suggestion
aws s3 mb $KOPS_STATE_STORE --region <region>
aws s3api put-bucket-versioning --bucket $KOPS_STATE_STORE --versioning-configuration Status=Enabled
```

I think this should enable versioning?

File: docs/howto/operate/add-aws-hub.md
Comment (yuvipanda) [inline docs/howto/operate/add-aws-hub.md:93]:
```suggestion
ssh-keygen -f <cluster-name>.key
```

This automatically creates the two, so you don't need the `mv`

File: docs/howto/operate/add-aws-hub.md
Comment (yuvipanda) [inline docs/howto/operate/add-aws-hub.md:122]: What do you think about not checking in the public key? We can always regenerate it from the private key with `ssh-keygen -y -f <cluster-name>.key`

File: docs/howto/operate/add-aws-hub.md
Comment (yuvipanda) [inline docs/howto/operate/add-aws-hub.md:207]: This needs to be repeated for staging and prod as well, no?

**Review by yuvipanda — CHANGES_REQUESTED** [review]
Minor changes but otherwise LGTM!

File: docs/howto/operate/add-aws-hub.md
Comment (damianavila) [reply docs/howto/operate/add-aws-hub.md:36]: > Does aws configure populate ~/.aws/credentials?

Yep, it does. In fact, those are already in the credentials file (this is why you can get them with `aws configure get`).
But AFAIK, when you export the profile name, the AWS_ACCESS_KEY_ID and AWS_SECRET_ACCESS_KEY env variables are not exported as well and kops seems to need those accordingly to the docs (check the last box at that section): https://kops.sigs.k8s.io/getting_started/aws/#setup-iam-user.

File: docs/howto/operate/add-aws-hub.md
Comment (damianavila) [reply docs/howto/operate/add-aws-hub.md:47]: Yep, it should be cluster instead of hub!

File: docs/howto/operate/add-aws-hub.md
Comment (damianavila) [reply docs/howto/operate/add-aws-hub.md:48]: Yep, I agree. That should be the command to enabled versioning.

File: docs/howto/operate/add-aws-hub.md
Comment (damianavila) [reply docs/howto/operate/add-aws-hub.md:123]: I agree, I will remove that line and make a note about how to recover it.
Btw, we should probably remove the public keys we have already pushed into the repo.

File: docs/howto/operate/add-aws-hub.md
Comment (damianavila) [reply docs/howto/operate/add-aws-hub.md:208]: Yes, I will make a note about that as well.

File: docs/howto/operate/add-aws-hub.md
Comment (damianavila) [reply docs/howto/operate/add-aws-hub.md:36]: Since this is minor, I will not block the merge on it.
We can always iterate and change it later down the road.
Btw, feel free to continue the discussion.

**Pull request merged**
